File: clients/typescript/src/coupled.ts
/**
 * Coupled system abstraction for host-parasite debugging.
 *
 * Manages both host and parasite as a single unit, hiding the complexity
 * of bus-stretch cancellation, pacing asymmetry, and stop ordering.
 */

import type { Beebium } from "./client.js";

export class CoupledSystem {
    private readonly host: Beebium;
    private readonly parasite: Beebium;
    private readonly ownsParasite: boolean;

    /**
     * Create a coupled system from existing host and parasite clients.
     *
     * @param host - The host BBC Micro client.
     * @param parasite - The parasite (second processor) client.
     * @param ownsParasite - If true, close the parasite on close().
     */
    constructor(host: Beebium, parasite: Beebium, ownsParasite = false) {
        this.host = host;
        this.parasite = parasite;
        this.ownsParasite = ownsParasite;
    }

    /**
     * Create a coupled system by discovering the parasite from the host.
     */
    static async fromHost(host: Beebium): Promise<CoupledSystem> {
        const parasite = await host.connectParasite();
        return new CoupledSystem(host, parasite, true);
    }

    /** Get the host client. */
    getHost(): Beebium {
        return this.host;
    }

    /** Get the parasite client. */
    getParasite(): Beebium {
        return this.parasite;
    }

    /** Run both processors. Ensures both are running afterwards. */
    async run(): Promise<void> {
        if (!await this.host.debugger.isRunning()) await this.host.debugger.run();
        if (!await this.parasite.debugger.isRunning()) await this.parasite.debugger.run();
    }

    /**
     * Stop both processors (bus-stretch safe).
     *
     * Stopping either side breaks the other out of any bus-stretch
     * spin-wait via the bus_stretch_cancel flag in TubeShared.
     */
    async stop(): Promise<void> {
        if (await this.host.debugger.isRunning()) await this.host.debugger.stop();
        if (await this.parasite.debugger.isRunning()) await this.parasite.debugger.stop();
    }

    /**
     * Run both processors until predicate returns true or the budget expires.
     *
     * Both processors advance at their natural rates. The predicate
     * is evaluated periodically via peek (side-effect-free) without
     * stopping either processor.
     *
     * @param predicate - Async function returning true when the condition is met.
     * @param emulatedSeconds - Maximum BBC-time seconds to run.
     * @param pollIntervalMs - Real-time milliseconds between predicate checks.
     * @returns true if the predicate was satisfied, false on timeout.
     */
    /**
     * Run both processors until predicate returns true or the budget expires.
     *
     * Execution proceeds in chunks of emulated time. At the end of each
     * chunk, both processors stop (via a server-side cycle-budget
     * breakpoint on the host with stopCounterpart), the predicate is
     * evaluated, and if false, both resume. No wall-clock polling.
     */
    async runUntil(
        predicate: () => Promise<boolean>,
        emulatedSeconds: number,
        chunkSeconds = 0.1,
    ): Promise<boolean> {
        const clockHz = await this.host.system.getClockSpeedHz() || 2_000_000;
        const totalBudget = Math.round(emulatedSeconds * clockHz);
        const chunkCycles = Math.round(chunkSeconds * clockHz);
        const startCycles = (await this.host.debugger.getState()).cycleCount;
        const deadlineCycles = startCycles + totalBudget;

        try {
            while ((await this.host.debugger.getState()).cycleCount < deadlineCycles) {
                const currentCycles = (await this.host.debugger.getState()).cycleCount;
                const chunkTarget = Math.min(currentCycles + chunkCycles, deadlineCycles);

                const bpId = await this.host.debugger.addBreakpoint(0x0000, {
                    endAddress: 0x10000,
                    condition: `cycles >= ${chunkTarget}`,
                    stopCounterpart: true,
                });

                // Ensure both stopped, subscribe to the stream, record
                // the sequence, THEN run. Any stopped event with a sequence
                // higher than recorded after run() is our breakpoint.
                await this.stop();
                const iter = this.host.debugger.watchExecutionState();
                const initial = await iter.next();
                const runSequence = initial.value!.state.sequence;
                await this.run();
                for await (const event of iter) {
                    if (!event.state.isRunning && event.state.sequence > runSequence) break;
                }

                if (await this.parasite.debugger.isRunning()) {
                    await this.parasite.debugger.stop();
                }
                await this.host.debugger.removeBreakpoint(bpId);

                if (await predicate()) {
                    return true;
                }
            }
            return predicate();
        } finally {
            await this.stop();
        }
    }

    /**
     * Run both processors for the given emulated time. No polling.
     */
    async runFor(emulatedSeconds: number): Promise<void> {
        await this.runUntil(async () => false, emulatedSeconds);
    }

    /**
     * Close the coupled system, stopping both processors.
     * If the parasite was auto-discovered, it is closed.
     */
    async close(): Promise<void> {
        await this.stop();
        if (this.ownsParasite) {
            await this.parasite.close();
        }
    }
}

File: clients/python/src/beebium/coupled.py
# Copyright 2026 Robert Smallshire <robert@smallshire.org.uk>
#
# This file is part of Beebium.
#
# Beebium is free software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version. Beebium is distributed in the hope that it will
# be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
# You should have received a copy of the GNU General Public License along with Beebium.
# If not, see <https://www.gnu.org/licenses/>.

"""Coupled system abstraction for host-parasite debugging.

Manages both host and parasite as a single unit, hiding the complexity
of bus-stretch cancellation, pacing asymmetry, and stop ordering.
"""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable
    from beebium.client import Beebium


class CoupledSystem:
    """Manages a host and parasite as a single unit.

    Both processors are run and stopped together. Predicates for
    ``run_until`` are evaluated periodically via peek (side-effect-free)
    without stopping either processor.

    Usage::

        system = CoupledSystem(host, parasite)
        system.run()
        # ...
        system.stop()

    Or with auto-discovery::

        system = CoupledSystem.from_host(bbc)
    """

    def __init__(self, host: Beebium, parasite: Beebium, *, owns_parasite: bool = False):
        """Create a coupled system from existing host and parasite clients.

        Args:
            host: The host BBC Micro client.
            parasite: The parasite (second processor) client.
            owns_parasite: If True, close the parasite on ``close()``.
        """
        self._host = host
        self._parasite = parasite
        self._owns_parasite = owns_parasite

    @classmethod
    def from_host(cls, host: Beebium, timeout: float = 5.0) -> CoupledSystem:
        """Create a coupled system by discovering the parasite from the host.

        Args:
            host: The host BBC Micro client.
            timeout: Connection timeout for the parasite.
        """
        parasite = host.connect_parasite(timeout=timeout)
        return cls(host, parasite, owns_parasite=True)

    @property
    def host(self) -> Beebium:
        return self._host

    @property
    def parasite(self) -> Beebium:
        return self._parasite

    def run(self) -> None:
        """Run both processors."""
        self._host.debugger.ensure_running()
        self._parasite.debugger.ensure_running()

    def stop(self) -> None:
        """Stop both processors (bus-stretch safe).

        Stopping either side breaks the other out of any bus-stretch
        spin-wait via the bus_stretch_cancel flag in TubeShared.
        """
        self._host.debugger.ensure_stopped()
        self._parasite.debugger.ensure_stopped()

    def run_until(
        self,
        predicate: Callable[[], bool],
        emulated_seconds: float,
        *,
        poll_interval: float = 0.02,
    ) -> bool:
        """Run both processors until predicate returns True or the budget expires.

        Both processors advance at their natural rates. The predicate
        is evaluated periodically via peek (side-effect-free) without
        stopping either processor. When the predicate fires or the
        budget is exceeded, both are stopped.

        Args:
            predicate: Callable returning True when the condition is met.
            emulated_seconds: Maximum BBC-time seconds to run.
            poll_interval: Real-time seconds between predicate checks.

        Returns:
            True if the predicate was satisfied, False on timeout.
        """
        clock_hz = self._host.system.clock_speed_hz or 2_000_000
        cycle_budget = int(emulated_seconds * clock_hz)
        start_cycles = self._host.debugger.cycle_count
        target_cycles = start_cycles + cycle_budget
        min_cycles_before_check = clock_hz  # 1 emulated second

        try:
            self.run()

            while True:
                time.sleep(poll_interval)
                current = self._host.debugger.cycle_count

                if current >= target_cycles:
                    self.stop()
                    return predicate()

                if current - start_cycles >= min_cycles_before_check:
                    if predicate():
                        self.stop()
                        return True
        except Exception:
            self.stop()
            raise

    def run_for(self, emulated_seconds: float) -> None:
        """Run both processors for the given emulated time.

        Uses a full-range breakpoint with a cycle condition on the host,
        with stop_counterpart to stop the parasite too. No polling.

        Args:
            emulated_seconds: BBC-time seconds to run.
        """
        clock_hz = self._host.system.clock_speed_hz or 2_000_000
        cycle_budget = int(emulated_seconds * clock_hz)
        target_cycles = self._host.debugger.cycle_count + cycle_budget

        bp_id = self._host.debugger.add_breakpoint(
            0x0000,
            end_address=0x10000,
            condition=f"cycles >= {target_cycles}",
            stop_counterpart=True,
        )
        try:
            stream = self._host.debugger.watch_execution_state()
            next(stream)  # consume initial state
            self.run()
            for event in stream:
                if not event.state.is_running:
                    break
        finally:
            self._host.debugger.remove_breakpoint(bp_id)
            self.stop()

    def close(self) -> None:
        """Close the coupled system, stopping both processors.

        If the parasite was auto-discovered (``owns_parasite=True``),
        it is closed. The host is not closed.
        """
        self.stop()
        if self._owns_parasite:
            self._parasite.close()
